### ISSUE TITLE
Compiler changed my var to const when using an anonymous struct.

### ISSUE THREAD

**Issue by taylorh140** [ziglang/zig#21388]
### Zig Version

0.13.0

### Steps to Reproduce and Observed Behavior

So i was debugging a variable in my code, because it was supposed to be incrementing. it wasn't.

I reduced the problem into this code which compiles: https://godbolt.org/z/886c9xWrn
```zig
const std = @import("std");

const Test_t = struct {
    a:usize = 0,
    b:usize,
};

pub var TestA = .{.b = 100};  //called @__anon_1427

export fn main() void{
    foo(&TestA);
}

fn foo(tst:*volatile Test_t) callconv(.C) void {
    tst.a+=1;
}
```

I thought this would add 1 to the TestA.a but it cant. because its constant!
It can be fixed by expliclty setting TestA's type:

```zig
const std = @import("std");

const Test_t = struct {
    a:usize = 0,
    b:usize,
};

pub var TestA = Test_t{.b = 100};  //called @example.TestA

export fn main() void{
    foo(&TestA);
}

fn foo(tst:*volatile Test_t) callconv(.C) void {
    tst.a+=1;
}
```

Looking at the LLVM IR:
```
before:
@__anon_1427 = internal unnamed_addr constant %example.Test_t { i64 0, i64 100 }, align 8

after:
@example.TestA = internal unnamed_addr global %example.Test_t { i64 0, i64 100 }, align 8
```

### Expected Behavior

I expected the variables to be placed into volatile memory like the .data section. However, in my case it was placed in .rodata and it kept my counter from counting.

**Comment by ikskuh** [comment]
`TestA` has an anonymous struct type that looks like this:
```zig
struct { comptime b: comptime_int =100 }
``|
when you take a pointer, it implicitly converts into a temporary of `Test_t`.

What you'e written is basically 
```zig
export fn main() void{
    foo(&.{ .b = 100 });
}
```

If you want your global to have an explicit type, the solution is to specify it like you in your second code or like this:
```zig
pub var TestA: Test_t = .{.b = 100};
```

**Comment by kibels** [comment]
> What you'e written is basically 
> ```zig
> export fn main() void{
>    foo(&.{ .b = 100 });
> }
> ```

Except that it was bound to the name of a var. The anonymous struct completely overriding that because its fields are comptime known is unexpected. Completely makes sense for literals but when you write

```zig
export fn main() void{
   var bar = .{ .b = 100 };
   foo(&bar);
 }
 ```
you expect the struct even if it's anonymous to be mutable. Should it at least be a compiler error that a const was passed to a function mutating it?

**Comment by mlugg** [comment]
The unexpectedly-allowed coercion here is solved by the accepted (but not yet implemented) proposal #16865.
